pkg/migration/migrations/fix_cast_for_style_migration.go

```diff
@@ -0,0 +1,154 @@
+	if err != nil {
+		return err
+	}
+
+	if len(alteredDesc) > 0 {
+		writeAllDesc(ctx, d, alteredDesc)
+	}
+
+	return nil
+}
+
+func filterAndFixDesc(
+	ctx context.Context,
+	semaCtx tree.SemaContext,
+	rows sqlutil.InternalRows,
+	visitor sql.FixCastForStyleVisitor,
+) ([]*descpb.TableDescriptor, error) {
+	var changedDesc []*descpb.TableDescriptor
+	ok, err := rows.Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	for ; ok; ok, err = rows.Next(ctx) {
+		datums := rows.Cur()
+		_, desc, _, err := unmarshalDescFromDescriptorRow(datums)
+		if err != nil {
+			return nil, err
+		}
+
+		tbDesc := desc.GetTable()
+		if tbDesc != nil {
+
+			columns := tbDesc.Columns
+			altered := false
+			for idx, col := range columns {
+				if col.IsComputed() {
+					computeExpr := *col.ComputeExpr
+					expr, err := parser.ParseExpr(computeExpr)
+					if err != nil {
+						return nil, err
+					}
+
+					visitor = sql.MakeFixCastForStyleVisitor(ctx, &semaCtx, tbDesc, col.Type, err)
+					newExpr, changed := tree.WalkExpr(&visitor, expr)
+					if changed {
+						altered = true
+						s := tree.Serialize(newExpr)
+						col.ComputeExpr = &s
+						columns[idx] = col
+					}
+				}
+			}
+
+			tIdx := tbDesc.Indexes
+			if len(tIdx) > 0 {
+				desc := tabledesc.NewBuilder(tbDesc)
+				tDesc := desc.BuildImmutableTable()
+
+				for idx, index := range tIdx {
+					if index.IsPartial() {
+						var typedExpr tree.TypedExpr
+						expr, err := parser.ParseExpr(index.Predicate)
+						replacedExpr, err := schemaexpr.MakeDummyColForTypeCheck(ctx, tDesc, expr, tree.NewUnqualifiedTableName(tree.Name(tDesc.GetName())))
+						//Are all/most partial expressions comparisonExpr?
+						if cExpr, ok := replacedExpr.(*tree.ComparisonExpr); ok {
+							typedExpr, err = cExpr.Left.TypeCheck(ctx, &semaCtx, types.Any)
+						} else {
+							typedExpr, err = replacedExpr.TypeCheck(ctx, &semaCtx, types.Any)
+						}
+
+						if err == nil {
+							typ := typedExpr.ResolvedType()
+							visitor = sql.MakeFixCastForStyleVisitor(ctx, &semaCtx, tbDesc, typ, err)
+							newExpr, changed := tree.WalkExpr(&visitor, expr)
+							if changed {
+								altered = true
+								s := tree.Serialize(newExpr)
+								index.Predicate = s
+								tIdx[idx] = index
+								tbDesc.Indexes = tIdx
+							}
+						}
+					}
+				}
+			}
+
+			if altered {
+				changedDesc = append(changedDesc, tbDesc)
+			}
+		}
+	}
+	return changedDesc, nil
+}
+
+func writeAllDesc(
+	ctx context.Context, d migration.TenantDeps, tDesc []*descpb.TableDescriptor,
+) error {
+	return d.CollectionFactory.Txn(ctx, d.InternalExecutor, d.DB, func(
+		ctx context.Context, txn *kv.Txn, descriptors *descs.Collection) error {
+		batch := txn.NewBatch()
+
+		for _, tDesc := range tDesc {
+			desc, err := descriptors.GetMutableDescriptorByID(ctx, tDesc.GetID(), txn)
+			if err != nil {
+				return err
+			}
+			mutDesc := desc.(*tabledesc.Mutable)
+			mutDesc.TableDescriptor = *tDesc
```

Comment:
This probably needs a look by schema, but unfortunately i'm not sure this is the best way to do it. Let me know if you need some peer programming time to tackle this together :) (or maybe someone else can help, given my jury duty). 

Things on my mind:
* We should be modifying the expressions using the mutable descriptor, instead of modifying it on the immutable descriptor and then assigning it to the mutable descriptor. otherwise, this can race - imagine someone writes a schema change between the time you read and modified the mutable descriptor, then you override it with your old view of the immutable descriptor.
* We should also probably tie the txn to only modify one descriptor at once.

These changes shouldn't be too different with what you have, just a little bit of rearrangement.

Copy pasting your code and putting that altogether, the pseudocode for this would probably look like:
```go
// step 1. get all tables. this replaces your query to fetch all descriptors manually
var tables []catalog.TableDescriptor

if err := ... d.CollectionFactory.Txn(ctx, d.InternalExecutor, d.DB, func(
		ctx context.Context, txn *kv.Txn, descriptors *descs.Collection) error {
   descs, err := descriptors. GetAllDescriptors(...)
   if err != nil {
      ...
   }

	for _, desc := range all.OrderedDescriptors() {
		if desc.GetParentID() == dbID {
			if table, ok := desc.(catalog.TableDescriptor); ok {
				tables = append(ret, table)
			}
		}
	}
}); err != nil {
 ...
}

// step 2. for each table, get mutable descriptor and fix
for _, table := range tables {
   if err := ... d.CollectionFactory.Txn(ctx, d.InternalExecutor, d.DB, func(
		ctx context.Context, txn *kv.Txn, descriptors *descs.Collection) error {

       desc, err := descriptors.GetMutableTableByID (. ... )
       // all your logic to fix everything up goes to `desc`.
       // call WriteDescToBatch
   }); err != nil {
       ...
   }
}
```